xdsl/utils/parse_pipeline.py

```diff
@@ -0,0 +1,227 @@
+    pipeline-element  ::= pass-name options?
+    options           ::= `{` options-element ( ` ` options-element)* `}`
+    options-element   ::= key (`=` value (`,` value)* )?
+
+    key       ::= IDENT
+    pass-name ::= IDENT
+    value     :== NUMBER / BOOL / IDENT / STRING_LITERAL
+    """
+
+    _stream: Iterator[Token]
+    _peeked: Token | None
+
+    def __init__(self, input_str: str):
+        self._stream = PipelineLexer._generator(input_str)
+        self._peeked = None
+
+    @staticmethod
+    def _generator(input_str: str) -> Iterator[Token]:
+        input = Input(input_str, "pass-pipeline")
+        pos = 0
+        end = len(input_str)
+
+        while True:
+            token: Token | None = None
+            for pattern, kind in _lexer_rules:
+                if (match := pattern.match(input_str, pos)) is not None:
+                    token = Token(Span(match.start(), match.end(), input), kind)
+                    pos = match.end()
+                    break
+            if token is None:
+                raise PassPipelineParseError(
+                    Token(Span(pos, pos + 1, input), Kind.IDENT), "Unknown token"
+                )
+            yield token
+            if pos >= end:
+                yield Token(Span(pos, pos + 1, input), Kind.EOF)
+                return
+
+    def lex(self) -> Token:
+        token = self.peek()
+        self._peeked = None
+        return token
+
+    def peek(self) -> Token:
+        if self._peeked is None:
+            self._peeked = next(self._stream)
+        return self._peeked
+
+
+class PassPipelineParseError(BaseException):
+    def __init__(self, token: Token, msg: str):
+        super().__init__(
+            "Error parsing pass pipeline specification:\n"
+            + token.span.print_with_context(msg)
+        )
+
+
+_PassArgTypes = list[str | int | bool | float]
+
+
+def parse_pipeline(
+    pipeline_spec: str,
+) -> Iterator[tuple[str, dict[str, _PassArgTypes]]]:
+    """
+    This takes a pipeline string and gives a representation of
+    the specification.
+
+    Each pass is represented by a tuple of:
+     - name: the name of the pass as string
+     - args: a dictionary, where each value is zero or more
+            of (str | bool | float | int)
+    """
+    lexer = PipelineLexer(pipeline_spec)
+
+    while True:
+        name = lexer.lex()
+        if name.kind is Kind.EOF:
+            return
+        if name.kind is not Kind.IDENT:
+            raise PassPipelineParseError(name, "Expected pass name here")
+
+        delim = lexer.lex()
+        if delim.kind is Kind.EOF:
+            yield name.span.text, dict()
+            return
+
+        if delim.kind is Kind.COMMA:
+            yield name.span.text, dict()
+            continue
+
+        if delim.kind is not Kind.L_BRACE:
+            raise PassPipelineParseError(
+                delim, "Expected a comma or pass arguments here"
+            )
```

Comment:
Here, to simplify the logic, you could instead just check for `L_BRACE`, and return the empty dictionary instead?
Here you are assuming that the only next tokens can be `EOF` or `COMMA`, which shouldn't be the concern of this function?

Reply:
I think I still want to raise an error in case the input is malformed, right? If someone makes an error an uses `pass1 pass2` it should tell it that the space there was unexpected.

Reply:
Well, you would get an error, though yes your error is `expected comma or pass arguments`, while the other way would give you `unexpected _`, so you would get a slightly worse error.
However, when maintaining code, this would mean that changing one rule might require to change other rules, as they reason about their position in the overall parser. I think this is the dangerous part, as it is very easy to miss that when you modify code, if you know what I mean.
Otherwise, I'm fine if you prefer it this way, I just find it a bit dangerous long-term.